File: src/lsp/hybrid.rs
//! FR-LSP-A: In-process hybrid typed call resolver (Go / TypeScript MVP).
//!
//! No child process, no JSON-RPC. Uses [`TypeRegistry`] built from the
//! current index pass to upgrade CALLS edges to `resolution_method=typed`.
//! Falls back silently when the registry cannot decide (ambiguous /
//! missing), leaving existing `name` / `unresolved` metadata intact.

use super::type_registry::{module_key_for_file, TypeRegistry};
use crate::config::typed_resolve_enabled;
use crate::db::models::Relationship;
use std::path::Path;

/// Result of a successful hybrid resolution.
#[derive(Debug, Clone, PartialEq)]
pub struct TypedHit {
    pub qualified_name: String,
    pub confidence: f64,
}

/// Resolve a bare callee name in the context of a calling file.
///
/// Order (mirrors CBM hybrid tiers at reduced fidelity):
/// 1. Same module/directory exact name
/// 2. Unique project-wide name
/// 3. Type.method when `receiver` is a known type
pub fn resolve_call(
    registry: &TypeRegistry,
    caller_file: &str,
    callee_name: &str,
    receiver: Option<&str>,
) -> Option<TypedHit> {
    if callee_name.is_empty() || callee_name.starts_with("__unresolved__") {
        let bare = callee_name.trim_start_matches("__unresolved__");
        return resolve_call(registry, caller_file, bare, receiver);
    }

    if let Some(rec) = receiver {
        if let Some(hit) = registry.lookup_type_method(rec, callee_name) {
            return Some(TypedHit {
                qualified_name: hit.qualified_name.clone(),
                confidence: 0.97,
            });
        }
    }

    let module = module_key_for_file(caller_file);
    if let Some(hit) = registry.lookup_in_module(&module, callee_name) {
        // Prefer a different file in the same module (cross-file), but
        // same-file hits still count as typed when the registry knows them.
        return Some(TypedHit {
            qualified_name: hit.qualified_name.clone(),
            confidence: if hit.file_path == caller_file {
                0.94
            } else {
                0.98
            },
        });
    }

    if let Some(hit) = registry.lookup_unique_name(callee_name) {
        return Some(TypedHit {
            qualified_name: hit.qualified_name.clone(),
            confidence: 0.92,
        });
    }

    None
}

/// Infer caller file path from a qualified name (`file::fn` or `file::Class::method`).
pub fn caller_file_from_qn(source_qualified: &str) -> String {
    let path = Path::new(source_qualified);
    // Prefer longest known file-like prefix before `::`
    if let Some(idx) = source_qualified.find("::") {
        let prefix = &source_qualified[..idx];
        if prefix.contains('.') || prefix.contains('/') {
            return prefix.to_string();
        }
    }
    path.to_string_lossy().to_string()
}

fn language_from_file(file_path: &str) -> Option<&'static str> {
    let ext = Path::new(file_path)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("");
    match ext {
        "go" => Some("go"),
        "ts" | "tsx" | "js" | "jsx" => Some("typescript"),
        // FR-B06: Python + Rust join the in-process hybrid resolver. The
        // registry indexes every function/class regardless of language, so
        // module (same-folder) and unique-name hits resolve Python defs and
        // Rust fns without spawning pylsp/rust-analyzer.
        "py" => Some("python"),
        "rs" => Some("rust"),
        "swift" => Some("swift"),
        "m" | "mm" => Some("objc"),
        // `.h` may be C or ObjC; hybrid resolve still runs when typed_resolve
        // includes objc and the index tagged the file as objc.
        "h" => Some("objc"),
        _ => None,
    }
}

fn bare_callee(target: &str) -> &str {
    if let Some(rest) = target.strip_prefix("__unresolved__") {
        return rest;
    }
    target.rsplit("::").next().unwrap_or(target)
}

fn receiver_from_metadata(meta: &serde_json::Value) -> Option<&str> {
    meta.get("receiver")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
}

/// Apply hybrid typed resolve to CALLS relationships in place.
///
/// Only touches edges whose source file language is enabled by
/// `typed_resolve` (Go / TypeScript / Swift / Objective-C). Returns the
/// number of edges upgraded to `resolution_method=typed`.
pub fn apply_typed_resolve(
    relationships: &mut [Relationship],
    registry: &TypeRegistry,
    typed_resolve: &str,
) -> usize {
    if registry.is_empty() {
        return 0;
    }
    let mut upgraded = 0;
    for rel in relationships.iter_mut() {
        if rel.rel_type != "calls" {
            continue;
        }
        let caller_file = caller_file_from_qn(&rel.source_qualified);
        let Some(lang) = language_from_file(&caller_file) else {
            continue;
        };
        if !typed_resolve_enabled(typed_resolve, lang) {
            continue;
        }

        // Already typed — leave alone.
        if rel
            .metadata
            .get("resolution_method")
            .and_then(|v| v.as_str())
            == Some("typed")
        {
            continue;
        }

        let callee = bare_callee(&rel.target_qualified);
        let receiver = receiver_from_metadata(&rel.metadata);
        let Some(hit) = resolve_call(registry, &caller_file, callee, receiver) else {
            continue;
        };

        // Do not downgrade a higher-confidence same-file name hit that
        // already points at the same QN — still mark typed.
        rel.target_qualified = hit.qualified_name;
        rel.confidence = hit.confidence.max(rel.confidence);
        let mut meta = rel.metadata.clone();
        if !meta.is_object() {
            meta = serde_json::json!({});
        }
        if let Some(obj) = meta.as_object_mut() {
            obj.insert("resolution_method".to_string(), serde_json::json!("typed"));
            obj.insert("is_resolved".to_string(), serde_json::json!(true));
            obj.insert("hybrid_tier".to_string(), serde_json::json!("in_process"));
        }
        rel.metadata = meta;
        upgraded += 1;
    }
    upgraded
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::models::CodeElement;
    use crate::lsp::type_registry::TypeRegistry;

    fn elem(qn: &str, etype: &str, name: &str, file: &str, lang: &str) -> CodeElement {
        CodeElement {
            qualified_name: qn.to_string(),
            element_type: etype.to_string(),
            name: name.to_string(),
            file_path: file.to_string(),
            line_start: 1,
            line_end: 2,
            language: lang.to_string(),
            parent_qualified: None,
            cluster_id: None,
            cluster_label: None,
            metadata: serde_json::json!({}),
            env: "local".to_string(),
        }
    }

    #[test]
    fn go_cross_file_same_package_is_typed() {
        let elements = vec![
            elem("pkg/a.go::Helper", "function", "Helper", "pkg/a.go", "go"),
            elem("pkg/b.go::Main", "function", "Main", "pkg/b.go", "go"),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let hit = resolve_call(&reg, "pkg/b.go", "Helper", None).unwrap();
        assert_eq!(hit.qualified_name, "pkg/a.go::Helper");
        assert!(hit.confidence >= 0.95);
    }

    #[test]
    fn typescript_unique_export_is_typed() {
        let elements = vec![elem(
            "src/util.ts::format",
            "function",
            "format",
            "src/util.ts",
            "typescript",
        )];
        let reg = TypeRegistry::from_elements(&elements);
        let hit = resolve_call(&reg, "src/app.ts", "format", None).unwrap();
        assert_eq!(hit.qualified_name, "src/util.ts::format");
    }

    #[test]
    fn python_same_module_def_is_typed() {
        // FR-B06: Python joins the in-process hybrid resolver via `py` ext.
        let elements = vec![
            elem(
                "pkg/utils.py::helper",
                "function",
                "helper",
                "pkg/utils.py",
                "python",
            ),
            elem(
                "pkg/app.py::main",
                "function",
                "main",
                "pkg/app.py",
                "python",
            ),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let hit = resolve_call(&reg, "pkg/app.py", "helper", None).unwrap();
        assert_eq!(hit.qualified_name, "pkg/utils.py::helper");
        assert!(hit.confidence >= 0.95);
    }

    #[test]
    fn rust_same_module_fn_is_typed() {
        // FR-B06: Rust joins via `rs` ext; unique-name cross-file also works.
        let elements = vec![
            elem(
                "src/math.rs::square",
                "function",
                "square",
                "src/math.rs",
                "rust",
            ),
            elem("src/main.rs::run", "function", "run", "src/main.rs", "rust"),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let hit = resolve_call(&reg, "src/main.rs", "square", None).unwrap();
        assert_eq!(hit.qualified_name, "src/math.rs::square");
        assert!(hit.confidence >= 0.95);
    }

    #[test]
    fn apply_typed_resolve_upgrades_python_and_rust_calls() {
        // FR-B06: `typed_resolve=all` must upgrade py/rs CALLS edges too.
        let elements = vec![
            elem(
                "pkg/utils.py::helper",
                "function",
                "helper",
                "pkg/utils.py",
                "python",
            ),
            elem(
                "pkg/app.py::main",
                "function",
                "main",
                "pkg/app.py",
                "python",
            ),
            elem(
                "src/math.rs::square",
                "function",
                "square",
                "src/math.rs",
                "rust",
            ),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let mut rels = vec![
            Relationship {
                id: None,
                source_qualified: "pkg/app.py::main".to_string(),
                target_qualified: "__unresolved__helper".to_string(),
                rel_type: "calls".to_string(),
                confidence: 0.5,
                metadata: serde_json::json!({"resolution_method": "unresolved"}),
                env: "local".to_string(),
            },
            Relationship {
                id: None,
                source_qualified: "src/main.rs::run".to_string(),
                target_qualified: "__unresolved__square".to_string(),
                rel_type: "calls".to_string(),
                confidence: 0.5,
                metadata: serde_json::json!({"resolution_method": "unresolved"}),
                env: "local".to_string(),
            },
        ];
        let n = apply_typed_resolve(&mut rels, &reg, "all");
        assert_eq!(n, 2, "python + rust CALLS both upgrade under all");
        for rel in &rels {
            assert_eq!(
                rel.metadata["resolution_method"].as_str(),
                Some("typed"),
                "{}",
                rel.source_qualified
            );
        }
        assert_eq!(
            rels[0].target_qualified, "pkg/utils.py::helper",
            "python target resolves"
        );
        assert_eq!(
            rels[1].target_qualified, "src/math.rs::square",
            "rust target resolves via unique name"
        );
    }

    #[test]
    fn apply_typed_resolve_upgrades_unresolved_calls() {
        let elements = vec![
            elem("pkg/a.go::Helper", "function", "Helper", "pkg/a.go", "go"),
            elem("pkg/b.go::Main", "function", "Main", "pkg/b.go", "go"),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let mut rels = vec![Relationship {
            id: None,
            source_qualified: "pkg/b.go::Main".to_string(),
            target_qualified: "__unresolved__Helper".to_string(),
            rel_type: "calls".to_string(),
            confidence: 0.5,
            metadata: serde_json::json!({"resolution_method": "unresolved"}),
            env: "local".to_string(),
        }];
        let n = apply_typed_resolve(&mut rels, &reg, "go,ts");
        assert_eq!(n, 1);
        assert_eq!(rels[0].target_qualified, "pkg/a.go::Helper");
        assert_eq!(
            rels[0].metadata["resolution_method"].as_str(),
            Some("typed")
        );
    }

    #[test]
    fn typed_resolve_off_skips_upgrade() {
        let elements = vec![elem(
            "pkg/a.go::Helper",
            "function",
            "Helper",
            "pkg/a.go",
            "go",
        )];
        let reg = TypeRegistry::from_elements(&elements);
        let mut rels = vec![Relationship {
            id: None,
            source_qualified: "pkg/b.go::Main".to_string(),
            target_qualified: "__unresolved__Helper".to_string(),
            rel_type: "calls".to_string(),
            confidence: 0.5,
            metadata: serde_json::json!({"resolution_method": "unresolved"}),
            env: "local".to_string(),
        }];
        assert_eq!(apply_typed_resolve(&mut rels, &reg, "off"), 0);
    }

    #[test]
    fn apply_typed_resolve_upgrades_swift_calls() {
        let elements = vec![
            elem(
                "App/Session.swift::authenticate",
                "method",
                "authenticate",
                "App/Session.swift",
                "swift",
            ),
            elem(
                "App/Session.swift::start",
                "method",
                "start",
                "App/Session.swift",
                "swift",
            ),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let mut rels = vec![Relationship {
            id: None,
            source_qualified: "App/Session.swift::start".to_string(),
            target_qualified: "App/Session.swift::authenticate".to_string(),
            rel_type: "calls".to_string(),
            confidence: 0.7,
            metadata: serde_json::json!({"resolution_method": "name"}),
            env: "local".to_string(),
        }];
        let n = apply_typed_resolve(&mut rels, &reg, "go,ts,swift,objc");
        assert_eq!(n, 1, "swift CALLS should upgrade when enabled");
        assert_eq!(
            rels[0].metadata["resolution_method"].as_str(),
            Some("typed")
        );
    }

    #[test]
    fn apply_typed_resolve_upgrades_objc_calls() {
        let elements = vec![
            elem("Greeter.m::setup", "method", "setup", "Greeter.m", "objc"),
            elem(
                "Greeter.m::sayHello",
                "method",
                "sayHello",
                "Greeter.m",
                "objc",
            ),
        ];
        let reg = TypeRegistry::from_elements(&elements);
        let mut rels = vec![Relationship {
            id: None,
            source_qualified: "Greeter.m::sayHello".to_string(),
            target_qualified: "Greeter.m::setup".to_string(),
            rel_type: "calls".to_string(),
            confidence: 0.7,
            metadata: serde_json::json!({"resolution_method": "name"}),
            env: "local".to_string(),
        }];
        let n = apply_typed_resolve(&mut rels, &reg, "swift,objc");
        assert_eq!(n, 1);
        assert_eq!(
            rels[0].metadata["resolution_method"].as_str(),
            Some("typed")
        );
    }

    #[test]
    fn never_spawns_process_on_resolve() {
        // Smoke: pure function path — no process handles involved.
        let reg = TypeRegistry::default();
        assert!(resolve_call(&reg, "x.go", "missing", None).is_none());
    }
}

File: go/internal/lsp/hybrid.go
package lsp

// In-process hybrid typed resolve (Rust parity: src/lsp/type_registry.rs +
// src/lsp/hybrid.rs, FR-LSP-A/C/D). No child process, no JSON-RPC: a symbol
// registry built from the indexed elements upgrades `calls` edges to
// resolution_method=typed, matching the Rust tiers (type.method, same module,
// unique project-wide name). Edges the registry cannot decide are left
// untouched.

import (
	"path"
	"strings"

	"github.com/FreePeak/LeanKG/go/internal/store"
)

// SymbolKind classifies one registry entry (Rust parity: SymbolKind).
type SymbolKind int

const (
	KindFunction SymbolKind = iota
	KindMethod
	KindType
)

// SymbolRef is one resolvable symbol (Rust parity: SymbolRef).
type SymbolRef struct {
	QualifiedName string
	FilePath      string
	Name          string
	Kind          SymbolKind
	TypeName      string // owning type/class/struct for methods
	Language      string
}

// TypeRegistry is the project-wide lookup table for hybrid typed resolve.
type TypeRegistry struct {
	byName       map[string][]SymbolRef
	byModuleName map[moduleName]SymbolRef
	byTypeMethod map[moduleName]SymbolRef
	fileModule   map[string]string
}

type moduleName struct{ Module, Name string }

// NewTypeRegistry builds the registry from indexed elements (Rust parity:
// TypeRegistry::from_elements). Methods/constructors are keyed by their
// parent's last qualified-name segment so Type.method lookups work.
func NewTypeRegistry(elements []store.Element) *TypeRegistry {
	r := &TypeRegistry{
		byName:       map[string][]SymbolRef{},
		byModuleName: map[moduleName]SymbolRef{},
		byTypeMethod: map[moduleName]SymbolRef{},
		fileModule:   map[string]string{},
	}
	for _, e := range elements {
		module := moduleKeyForFile(e.FilePath)
		r.fileModule[e.FilePath] = module
		switch e.ElementType {
		case "function":
			r.insertSymbol(module, SymbolRef{
				QualifiedName: e.QualifiedName, FilePath: e.FilePath, Name: e.Name,
				Kind: KindFunction, Language: e.Language,
			})
		case "method", "constructor":
			typeName := ""
			if e.ParentQualified != "" && e.ParentQualified != e.FilePath {
				if i := strings.LastIndex(e.ParentQualified, "::"); i >= 0 {
					typeName = e.ParentQualified[i+2:]
				}
			}
			sym := SymbolRef{
				QualifiedName: e.QualifiedName, FilePath: e.FilePath, Name: e.Name,
				Kind: KindMethod, TypeName: typeName, Language: e.Language,
			}
			if typeName != "" {
				r.byTypeMethod[moduleName{typeName, e.Name}] = sym
			}
			r.insertSymbol(module, sym)
		case "class", "struct", "interface", "type", "enum":
			r.insertSymbol(module, SymbolRef{
				QualifiedName: e.QualifiedName, FilePath: e.FilePath, Name: e.Name,
				Kind: KindType, Language: e.Language,
			})
		}
	}
	return r
}

func (r *TypeRegistry) insertSymbol(module string, sym SymbolRef) {
	r.byModuleName[moduleName{module, sym.Name}] = sym
	r.byName[sym.Name] = append(r.byName[sym.Name], sym)
}

// ModuleForFile returns the module key recorded for a file path.
func (r *TypeRegistry) ModuleForFile(filePath string) (string, bool) {
	m, ok := r.fileModule[filePath]
	return m, ok
}

// LookupInModule is the exact (module, name) hit — preferred for same-package
// Go / same-folder TS.
func (r *TypeRegistry) LookupInModule(module, name string) (SymbolRef, bool) {
	s, ok := r.byModuleName[moduleName{module, name}]
	return s, ok
}

// LookupUniqueName is the project-wide match when exactly one candidate
// exists.
func (r *TypeRegistry) LookupUniqueName(name string) (SymbolRef, bool) {
	hits := r.byName[name]
	if len(hits) == 1 {
		return hits[0], true
	}
	return SymbolRef{}, false
}

// LookupTypeMethod resolves Type.method on a known type.
func (r *TypeRegistry) LookupTypeMethod(typeName, method string) (SymbolRef, bool) {
	s, ok := r.byTypeMethod[moduleName{typeName, method}]
	return s, ok
}

// Candidates returns every symbol sharing a bare name.
func (r *TypeRegistry) Candidates(name string) []SymbolRef { return r.byName[name] }

// Len returns the number of registered symbols.
func (r *TypeRegistry) Len() int {
	n := 0
	for _, v := range r.byName {
		n += len(v)
	}
	return n
}

// IsEmpty reports whether nothing was registered.
func (r *TypeRegistry) IsEmpty() bool { return len(r.byName) == 0 }

// moduleKeyForFile is the directory containing the file ("." at the root),
// used as the lightweight module key when no package/import is parsed (Rust
// parity: module_key_for_file).
func moduleKeyForFile(filePath string) string {
	d := path.Dir(strings.ReplaceAll(filePath, "\\", "/"))
	if d == "" || d == "." || d == "/" {
		return "."
	}
	return d
}

// TypedHit is a successful hybrid resolution (Rust parity: TypedHit).
type TypedHit struct {
	QualifiedName string
	Confidence    float64
}

// ResolveCall resolves a bare callee name in the context of a calling file,
// in Rust tier order: receiver type method (0.97), same module (0.94 same
// file / 0.98 cross file), unique project-wide name (0.92).
func ResolveCall(r *TypeRegistry, callerFile, calleeName, receiver string) (TypedHit, bool) {
	if calleeName == "" {
		return TypedHit{}, false
	}
	if rest, ok := strings.CutPrefix(calleeName, "__unresolved__"); ok {
		return ResolveCall(r, callerFile, rest, receiver)
	}
	if receiver != "" {
		if hit, ok := r.LookupTypeMethod(receiver, calleeName); ok {
			return TypedHit{hit.QualifiedName, 0.97}, true
		}
	}
	module := moduleKeyForFile(callerFile)
	if hit, ok := r.LookupInModule(module, calleeName); ok {
		conf := 0.98
		if hit.FilePath == callerFile {
			conf = 0.94
		}
		return TypedHit{hit.QualifiedName, conf}, true
	}
	if hit, ok := r.LookupUniqueName(calleeName); ok {
		return TypedHit{hit.QualifiedName, 0.92}, true
	}
	return TypedHit{}, false
}

// CallerFileFromQN infers the caller file from a qualified name
// (`file::fn` / `file::Type.method`, Rust parity: caller_file_from_qn).
func CallerFileFromQN(sourceQualified string) string {
	if i := strings.Index(sourceQualified, "::"); i > 0 {
		prefix := sourceQualified[:i]
		if strings.ContainsAny(prefix, "./") {
			return prefix
		}
	}
	return sourceQualified
}

// LanguageFromFile maps a caller file path to the hybrid-resolve language id
// (Rust parity: hybrid.rs language_from_file — go / typescript / swift /
// objc only).
func LanguageFromFile(filePath string) (string, bool) {
	switch strings.ToLower(path.Ext(filePath)) {
	case ".go":
		return "go", true
	case ".ts", ".tsx", ".js", ".jsx":
		return "typescript", true
	case ".swift":
		return "swift", true
	case ".m", ".mm", ".h":
		return "objc", true
	}
	return "", false
}

// typedResolveAliases mirrors the Rust alias table: canonical -> aliases.
var typedResolveAliases = [][2]string{
	{"typescript", "ts,tsx,typescript,javascript,js,jsx"},
	{"javascript", "js,jsx,javascript"},
	{"python", "py,python"},
	{"rust", "rs,rust"},
	{"ruby", "rb,ruby"},
	{"csharp", "cs,csharp,c#"},
	{"swift", "swift"},
	{"objc", "objc,objective-c,objectivec,m,mm"},
}

// TypedResolveEnabled interprets the typed_resolve setting for one language
// (Rust parity: config::project::typed_resolve_enabled). Accepted values:
// off/""/false/no, all/true/yes/on, or a CSV of language names/aliases.
func TypedResolveEnabled(setting, language string) bool {
	s := strings.ToLower(strings.TrimSpace(setting))
	switch s {
	case "off", "", "false", "no":
		return false
	case "all", "true", "yes", "on":
		return true
	}
	langLower := strings.ToLower(language)
	accepted := map[string]bool{langLower: true}
	for _, a := range typedResolveAliases {
		canonical, list := a[0], strings.Split(a[1], ",")
		if contains(list, langLower) {
			accepted[canonical] = true
		}
		if canonical == langLower {
			for _, alias := range list {
				accepted[alias] = true
			}
		}
	}
	for _, part := range strings.FieldsFunc(s, func(r rune) bool {
		return r == ',' || r == ' ' || r == ';'
	}) {
		if accepted[part] || part == "all" {
			return true
		}
	}
	return false
}

func contains(list []string, s string) bool {
	for _, v := range list {
		if v == s {
			return true
		}
	}
	return false
}

// ApplyTypedResolve upgrades CALLS edges in place (Rust parity:
// apply_typed_resolve). Only edges whose source file language is enabled by
// the typed_resolve setting are touched; already-typed edges are skipped.
// Returns the number of upgraded edges.
func ApplyTypedResolve(rels []store.Relationship, r *TypeRegistry, setting string) int {
	n, _ := applyTypedResolveIdx(rels, r, setting)
	return n
}

// applyTypedResolveIdx is ApplyTypedResolve plus the indices it rewrote (the
// store pass writes back exactly those rows).
func applyTypedResolveIdx(rels []store.Relationship, r *TypeRegistry, setting string) (int, []int) {
	if r.IsEmpty() {
		return 0, nil
	}
	var idx []int
	for i := range rels {
		rel := &rels[i]
		if rel.RelType != "calls" {
			continue
		}
		callerFile := CallerFileFromQN(rel.Source)
		lang, ok := LanguageFromFile(callerFile)
		if !ok || !TypedResolveEnabled(setting, lang) {
			continue
		}
		if method, _ := rel.Metadata["resolution_method"].(string); method == "typed" {
			continue
		}
		receiver, _ := rel.Metadata["receiver"].(string)
		hit, ok := ResolveCall(r, callerFile, BareCallee(rel.Target), receiver)
		if !ok {
			continue
		}
		rel.Target = hit.QualifiedName
		if hit.Confidence > rel.Confidence {
			rel.Confidence = hit.Confidence
		}
		if rel.Metadata == nil {
			rel.Metadata = map[string]any{}
		}
		rel.Metadata["resolution_method"] = "typed"
		rel.Metadata["is_resolved"] = true
		rel.Metadata["hybrid_tier"] = "in_process"
		idx = append(idx, i)
	}
	return len(idx), idx
}

// BareCallee strips the __unresolved__ marker, any receiver prefix
// (`file::`) and the Go method type prefix (`Type.method`, the engine's
// method qualified-name shape). Rust parity: hybrid.rs bare_callee plus the
// Go QN separator.
func BareCallee(target string) string {
	if rest, ok := strings.CutPrefix(target, "__unresolved__"); ok {
		target = rest
	}
	if i := strings.LastIndex(target, "::"); i >= 0 {
		target = target[i+2:]
	}
	if i := strings.LastIndexByte(target, '.'); i >= 0 {
		target = target[i+1:]
	}
	return target
}

// HybridResolve runs the in-process typed resolve over the whole store:
// elements build the registry, every `calls` edge is upgraded in place and
// only the upgraded rows are written back. It spawns nothing.
func HybridResolve(st store.Backend, setting string) (int, error) {
	if setting == "" {
		setting = "off"
	}
	els, err := st.Elements()
	if err != nil {
		return 0, err
	}
	reg := NewTypeRegistry(els)
	if reg.IsEmpty() {
		return 0, nil
	}
	rels, err := st.RelationshipsAll(allRelationshipsLimit)
	if err != nil {
		return 0, err
	}
	n, idx := applyTypedResolveIdx(rels, reg, setting)
	if n == 0 {
		return 0, nil
	}
	changed := make([]store.Relationship, 0, n)
	for _, i := range idx {
		changed = append(changed, rels[i])
	}
	if err := st.UpsertRelationships(changed); err != nil {
		return 0, err
	}
	return n, nil
}

// allRelationshipsLimit bounds the store scan for hybrid resolve; the store's
// own default (1000) is too small for real projects.
const allRelationshipsLimit = 1 << 30
